parser/charset/encoding_table.go

```diff
@@ -298,3 +300,177 @@ func characterLengthUTF8(bs []byte) int {
+//   - TruncateStrategyTrim: returns the valid prefix part of string.
+//   - TruncateStrategyReplace: returns the whole string, but the invalid characters are replaced with '?'.
+type TruncateStrategy int8
+
+const (
+	TruncateStrategyEmpty TruncateStrategy = iota
+	TruncateStrategyTrim
+	TruncateStrategyReplace
+)
+
+var _ StringValidator = StringValidatorASCII{}
+var _ StringValidator = StringValidatorUTF8{}
+var _ StringValidator = StringValidatorOther{}
+
+// StringValidator is used to check if a string is valid in the specific charset.
+type StringValidator interface {
+	Validate(str string) (invalidPos int)
+	Truncate(str string, strategy TruncateStrategy) (result string, invalidPos int)
+}
+
+// StringValidatorASCII checks whether a string is valid ASCII string.
+type StringValidatorASCII struct{}
+
+// Validate checks whether the string is valid in the given charset.
+func (s StringValidatorASCII) Validate(str string) int {
+	_, invalidPos := s.Truncate(str, TruncateStrategyEmpty)
+	return invalidPos
+}
+
+// Truncate implement the interface StringValidator.
+func (s StringValidatorASCII) Truncate(str string, strategy TruncateStrategy) (string, int) {
+	invalidPos := -1
+	for i := 0; i < len(str); i++ {
+		if str[i] > go_unicode.MaxASCII {
+			invalidPos = i
+		}
+	}
+	if invalidPos == -1 {
+		// Quick check passed.
+		return str, -1
+	}
+	switch strategy {
+	case TruncateStrategyEmpty:
+		return "", invalidPos
+	case TruncateStrategyTrim:
+		return str[:invalidPos], invalidPos
+	case TruncateStrategyReplace:
+		result := make([]byte, 0, len(str))
+		for i, w := 0, 0; i < len(str); i += w {
+			w = 1
+			if str[i] > go_unicode.MaxASCII {
+				w = characterLengthUTF8(Slice(str)[i:])
+				w = mathutil.Min(w, len(str)-i)
+				result = append(result, '?')
+				continue
+			}
+			result = append(result, str[i:i+w]...)
+		}
+		return string(result), invalidPos
+	}
+	return str, -1
+}
+
+// StringValidatorUTF8 checks whether a string is valid UTF8 string.
+type StringValidatorUTF8 struct {
+	IsUTF8MB4           bool // Distinguish between "utf8" and "utf8mb4"
+	CheckMB4ValueInUTF8 bool
+}
+
+// Validate checks whether the string is valid in the given charset.
+func (s StringValidatorUTF8) Validate(str string) int {
+	_, invalidPos := s.Truncate(str, TruncateStrategyEmpty)
+	return invalidPos
+}
+
+// Truncate implement the interface StringValidator.
+func (s StringValidatorUTF8) Truncate(str string, strategy TruncateStrategy) (string, int) {
+	if s.IsUTF8MB4 && utf8.ValidString(str) {
+		// Quick check passed.
+		return str, -1
+	}
+	doMB4CharCheck := !s.IsUTF8MB4 && s.CheckMB4ValueInUTF8
+	var result []byte
+	if strategy == TruncateStrategyReplace {
+		result = make([]byte, 0, len(str))
+	}
+	invalidPos := -1
+	for i, w := 0, 0; i < len(str); i += w {
+		var rv rune
+		rv, w = utf8.DecodeRuneInString(str[i:])
+		if (rv == utf8.RuneError && !strings.HasPrefix(str[i:], string(utf8.RuneError))) ||
+			w > 3 && doMB4CharCheck {
```

Comment:
```suggestion
			(w > 3 && doMB4CharCheck) {
```
to make it more clear